### PR TITLE
add support for --fix-broken-easyconfigs + style fixes

### PR DESCRIPTION
**Pull request by boegel**
The intention of `--fix-broken-easyconfigs` is to automagically rewrite easyconfigs that were not adjusted yet to adher to the deprecated behaviour that is no longer supported in EasyBuild v2.0 (http://easybuild.readthedocs.org/en/latest/Deprecated-functionality.html#easyconfig-parameters), i.e.:

* replacing `(pre)makeopts` with `(pre)buildopts`
* replacing `shared_lib_ext` with `SHLIB_EXT`
* replacing `license` with `license_file`
* disabling the automagic fallback to the `ConfigureMake` easyblock, for software for which no dedicated easyblock is available

This avoids the need to manually apply (simple) fixes to make `eb` accept the easyconfigs again.

### CODE REVIEW

**Comment by boegel** [comment]
@stdweird: please review? I'd *love* to get this into EasyBuild v2.0.

**Comment by stdweird** [comment]
what does this try to solve? i can understand the inconvenience, but why is this fixed within EB? maybe when deprecating things, provide a cleanup wrapper script so people can convert broken ones to working ones. but keep that out of framework. this way, you are still keeping legacy things in framework.
if you really insist of doing it like this, then each new deprecated option must get the code to fix it etc etc etc.

i'm not happy with the passing `rawtxt` around hack.
i also don't understand how this code can format independent.

**Comment by boegel** [comment]
I'm more in favour of supporting this in the framework, because I want to avoid fragmentation and end up with a dozen different tools that provide all kinds of different functionality left and right... By supporting it via `eb` directly, people can figure it out via `eb --help`, and we can even mention this option in the error message when we notice old easyconfig parameters are still being relied upon (it's not doing that yet).

If you (and others) feel strongly about this, I'm willing to flesh this out in stand-alone script instead, to minimalise support for deprecated/broken stuff in the framework. We will still need some notion of the replaced parameters in the framework though, to do decent error reporting, but it would be limited to only a couple of lines of code.

I just feel there should be something available to make the transition easier, since lots of people will be sitting on a pile of easyconfig files that no longer work in EasyBuild 2.0 due to (mostly) stylistic non-backward-compatible changes... Not providing something will result in people having to script their own thing to avoid repetitive simple substitutions in the easyconfig files they have. I think we both feel that wouldn't be a good situation.

I don't see passing `rawtxt` passing around as a hack, it's an optimisation. I noticed that the framework was reading the easyconfig file several times, this is now significantly less so. It's kind of stupid to reread the easyconfig file from disk over and over, since it's not being changed (and shouldn't be changed from the outside while `eb` is running either).

This code is not meant to be format-independent; easyconfig format v2 is still experimental, so I'm not trying to autofix those; we haven't landed yet on the final format there, so we can get away with simply not supporting `makeopts` etc. there in the final format. That does mean that the code supporting this magic should be in the package specific to easyconfig format v1, I guess. Point taken.

**Comment by boegel** [comment]
ready to be re-reviewed... @wpoely86: up for it?

File: easybuild/framework/easyconfig/easyconfig.py [hunk 1]
```diff
@@ -123,9 +112,18 @@ def __init__(self, path, extra_options=None, build_specs=None, validate=True, hi
...
 
-        if not os.path.isfile(path):
+        if path is not None and not os.path.isfile(path):
             self.log.error("EasyConfig __init__ expected a valid path")
 
+        # read easyconfig file contents (or use provided rawtxt), so it can be passed down to avoid multiple re-reads
+        self.path = path
+        if rawtxt is None:
+            self.rawtxt = read_file(path)
+            self.log.info("Raw contents from supplied easyconfig file %s: %s" % (path, self.rawtxt))
```
Comment (stdweird) [inline easybuild/framework/easyconfig/easyconfig.py]: `debug` is enough here i guess (and also 3 lines below)

File: easybuild/framework/easyconfig/easyconfig.py [hunk 2]
```diff
@@ -123,9 +112,18 @@ def __init__(self, path, extra_options=None, build_specs=None, validate=True, hi
...
+        else:
+            self.rawtxt = rawtxt
+            self.log.info("Supplied easyconfig: %s" % self.rawtxt)
```
Comment (stdweird) [inline easybuild/framework/easyconfig/easyconfig.py]: also add `Raw contents` in the message (somehow)

File: easybuild/framework/easyconfig/easyconfig.py [hunk 3]
```diff
@@ -109,7 +98,7 @@ class EasyConfig(object):
     Class which handles loading, reading, validation of easyconfigs
     """
 
-    def __init__(self, path, extra_options=None, build_specs=None, validate=True, hidden=None):
+    def __init__(self, path, extra_options=None, build_specs=None, validate=True, hidden=None, rawtxt=None):
         """
         initialize an easyconfig.
         @param path: path to easyconfig file to be parsed
```
Comment (stdweird) [inline easybuild/framework/easyconfig/easyconfig.py]: i like how you add arguments without updating the docstring.
but there is something else really wrong if you need this many named args.

File: easybuild/framework/easyconfig/easyconfig.py [hunk 4]
```diff
@@ -692,16 +691,14 @@ def det_installversion(version, toolchain_name, toolchain_version, prefix, suffi
...
-    else:
-        return None
+    """
+    Fetch parameter specification from given easyconfig file.
+    DEPRECATED: use fetch_parameters_from_easyconfig from easybuild.framework.easyconfigs.parser instead
+    """
+    old = 'fetch_parameter_from_easyconfig_file'
+    new = 'fetch_parameters_from_easyconfig'
+    _log.deprecated("%s is replaced by %s from easybuild.framework.easyconfig.parser" % (old, new), '3.0')
+    return fetch_parameters_from_easyconfig(read_file(path), [param])[0]
```
Comment (stdweird) [inline easybuild/framework/easyconfig/easyconfig.py]: don't read the file in the function call. it's gives messed up stacktraces i think on error

File: easybuild/framework/easyconfig/easyconfig.py [hunk 5]
```diff
@@ -231,16 +229,17 @@ def parse(self):
             self.log.error("Specifications should be specified using a dictionary, got %s" % type(self.build_specs))
         self.log.debug("Obtained specs dict %s" % arg_specs)
 
-        parser = EasyConfigParser(self.path)
+        self.log.info("Parsing easyconfig file %s with rawcontent: %s" % (self.path, self.rawtxt))
+        parser = EasyConfigParser(filename=self.path, rawcontent=self.rawtxt)
```
Comment (stdweird) [inline easybuild/framework/easyconfig/easyconfig.py]: only if `self.path` is `None` is the rawcontent ever used in the `EasyConfigParser`? is this really intended?

File: easybuild/scripts/fix_broken_easyconfigs.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+import os
```
Comment (stdweird) [inline easybuild/scripts/fix_broken_easyconfigs.py]: add author

File: easybuild/scripts/fix_broken_easyconfigs.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.framework.easyconfig.easyconfig import get_easyblock_class
+from easybuild.framework.easyconfig.parser import REPLACED_PARAMETERS, fetch_parameters_from_easyconfig
+from easybuild.tools.filetools import find_easyconfigs, read_file, write_file
+
+from test.framework.utilities import init_config
+
+
+log = fancylogger.getLogger('main')
```
Comment (stdweird) [inline easybuild/scripts/fix_broken_easyconfigs.py]: use the `go.log` from simple_option?

File: easybuild/scripts/fix_broken_easyconfigs.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            log.error("Non-existing path %s specified" % path)
+
+    ec_files = [ec for p in go.args for ec in find_easyconfigs(p)]
+    if not ec_files:
+        log.error("No easyconfig files specified")
+
+    log.info("Processing %d easyconfigs" % len(ec_files))
+    for ec_file in ec_files:
+
+        ectxt = read_file(ec_file)
```
Comment (stdweird) [inline easybuild/scripts/fix_broken_easyconfigs.py]: move this in function instead of main

**Comment by stdweird** [comment]
@boegel looks ok, except for the `EasyConfigParser` issue

File: easybuild/framework/easyconfig/easyconfig.py (hunk 5 above)
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: should be the other way around, you're right

I'll fix that, and add a unit test that checks for this behaviour

**Comment by boegel** [comment]
remarks fixed, unit tests are happy, going in!

**Comment by boegel** [comment]
thanks @stdweird for the review!

**Pull request merged**